runtime/runtime_storage.go

```diff
@@ -211,41 +212,122 @@ func (s *interpreterRuntimeStorage) writeValue(
 //
 func (s *interpreterRuntimeStorage) writeCached() {
 
+	type writeItem struct {
+		storageKey storageKey
+		value      interpreter.Value
+	}
+
+	var items []writeItem
+
 	for fullKey, entry := range s.cache {
 
 		if !entry.mustWrite && entry.value != nil && !entry.value.Modified() {
 			continue
 		}
 
+		items = append(items, writeItem{
+			storageKey: fullKey,
+			value:      entry.value,
+		})
+	}
+
+	// Don't use a for-range loop, as keys are added while iterating
+	for len(items) > 0 {
```

Comment:
I wonder about pulling this out into a separate function, might make it easier to unit test down the road -- `determineDefferals` or something perhaps.

Reply:
I like splitting this up, but the `determineDeferrals` also then encodes values? Maybe `encodeAndDetermineDeferrals` is a better name?